### PR TITLE
Dev twa mops

### PR DESCRIPTION
**Pull request by jb2197**
This PR implements OntoMOPs and its associated algorithms using twa OGM.

### CODE REVIEW

File: MOPTools/MOP_Discovery/main.py [hunk 1]
```diff
@@ -22,12 +22,12 @@ def mops_output():
...
-    os.mkdir(FOLDER_NAME+'\\r2_cbus')
-    os.mkdir(FOLDER_NAME+'\\temp')
-    os.mkdir(FOLDER_NAME+'\\mops_r1')
-    os.mkdir(FOLDER_NAME+'\\mops_r2')
+    os.mkdir(FOLDER_NAME+'/mops_am')
+    os.mkdir(FOLDER_NAME+'/r1_cbus')
+    os.mkdir(FOLDER_NAME+'/r2_cbus')
+    os.mkdir(FOLDER_NAME+'/temp')
+    os.mkdir(FOLDER_NAME+'/mops_r1')
+    os.mkdir(FOLDER_NAME+'/mops_r2')
```
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/MOP_Discovery/main.py]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/aa7ca002131b186522290cd1d8dcc039371906d9

File: MOPTools/MOP_Discovery/query_operations/queryendpoint.py
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/MOP_Discovery/query_operations/queryendpoint.py:1]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/589f6ad5e68cac971f853a9840929bf5d481e702

File: MOPTools/twa_mops/alg1.py
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/twa_mops/alg1.py:1]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/f9373ade2923bf44785a29c497e94ce8a0fbff0e

File: MOPTools/twa_mops/alg2.py
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/twa_mops/alg2.py:1]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/f9373ade2923bf44785a29c497e94ce8a0fbff0e

File: MOPTools/MOP_Discovery/manager/file_paths.py [hunk 2]
```diff
@@ -3,19 +3,21 @@
...
+    'kg_assembly_csv': os.path.join(base, 'mops_output/kg_analysis_1.csv'),
+    'mops_am': os.path.join(base, 'mops_output/mops_am/'),
+    'r1_cbus': os.path.join(base, 'mops_output/r1_cbus/'),
+    'r2_cbus': os.path.join(base, 'mops_output/r2_cbus/'),
+    'list_R1': os.path.join(base, 'mops_output/list_R1.json'),
+    'list_preR2': os.path.join(base, 'mops_output/list_preR2.json'),
+    'list_R2': os.path.join(base, 'mops_output/list_R2.json'),
+    'temp': os.path.join(base, 'mops_output/temp/'),
+    'mops_r1': os.path.join(base, 'mops_output/mops_r1/'),
+    'mops_r2': os.path.join(base, 'mops_output/mops_r2/')
```
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/MOP_Discovery/manager/file_paths.py]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/commit/aa7ca002131b186522290cd1d8dcc039371906d9

File: MOPTools/twa_mops/main.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+from twa.kg_operations import PySparqlClient
+from rdflib import Graph, URIRef
+import ontomops
+import ontospecies
+import om
+import alg2
+
+endpoint = 'http://localhost:48082/blazegraph/namespace/ontomops151/sparql'
```
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/twa_mops/main.py]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/4aa9530366d6f3ca52f591bfc1ed131385fc2807

File: MOPTools/twa_mops/ontomops_backup.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+# Get the current date in YYYY-MM-DD format
+current_date=$(date +"%Y-%m-%d")
+
+# Define the output file name using the current date
+output_file="ontomops_backup_${current_date}.ttl"
+
+# Execute the curl command and save the output to the specified file
+curl -X POST --url 'http://localhost:48082/blazegraph/namespace/ontomops151/sparql' \
```
[the remark this answers is not in the document]
Comment (jb2197) [reply MOPTools/twa_mops/ontomops_backup.sh]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/4aa9530366d6f3ca52f591bfc1ed131385fc2807

File: MOPTools/twa_mops/ontospecies.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+from __future__ import annotations
+from typing import List, Optional
+from twa.data_model.base_ontology import BaseOntology, BaseClass, ObjectProperty, DatatypeProperty
+import om
+import geo
+
+from rdkit.Chem import GetPeriodicTable
+from rdkit.Chem.rdmolfiles import MolFromXYZFile
+
+# NOTE TODO this script is incomplete
```
Comment (sm453) [inline MOPTools/twa_mops/ontospecies.py:10]: This isn't a very helpful to-do note. Please explain what is missing.

File: MOPTools/twa_mops/om.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+from __future__ import annotations
+from twa.data_model.base_ontology import BaseOntology, BaseClass, ObjectProperty, DatatypeProperty
+
+
+# NOTE TODO this script is incomplete, and it should be moved to a place that accessible to all other ontologies
```
Comment (sm453) [inline MOPTools/twa_mops/om.py:5]: Please briefly indicate what is missing.

File: MOPTools/MOP_Discovery/query_operations/config.json [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{
+    "ontospecies": "http://localhost:48082/blazegraph/namespace/ontomops/sparql",
+    "ontomops": "http://localhost:48082/blazegraph/namespace/ontomops/sparql"
```
Comment (sm453) [inline MOPTools/MOP_Discovery/query_operations/config.json:3]: Are these two meant to be identical, i.e. the species one also pointing to the mops endpoint?

File: MOPTools/MOP_Discovery/query_operations/queryendpoint.py [hunk 8]
```diff
@@ -1,4 +1,15 @@
-SPARQL_ENDPOINTS = {
-    'ontospecies': 'http://theworldavatar.com/blazegraph/namespace/ontospecies/sparql',
-    'ontomops': 'https://kg.cmclinnovations.com/blazegraph_geo/namespace/ontomops/sparql'
```
Comment (sm453) [inline MOPTools/MOP_Discovery/query_operations/queryendpoint.py:3]: This does not look like what you put into the config file...

**Review by sm453 — CHANGES_REQUESTED** [review]

File: MOPTools/MOP_Discovery/query_operations/config.json (hunk 7 above)
Comment (jb2197) [reply MOPTools/MOP_Discovery/query_operations/config.json:3]: Yes, they are meant to be the same endpoint following the refactoring of OntoMOPs TBox and ABox design.

File: MOPTools/MOP_Discovery/query_operations/queryendpoint.py (hunk 8 above)
Comment (jb2197) [reply MOPTools/MOP_Discovery/query_operations/queryendpoint.py:3]: As discussed, these endpoints are no longer working and are replaced with the ones in the config.

File: MOPTools/twa_mops/om.py (hunk 6 above)
Comment (jb2197) [reply MOPTools/twa_mops/om.py]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/2b36b35b9b6d806f9bbabfc6bebf16ee822f2f51

File: MOPTools/twa_mops/ontospecies.py (hunk 5 above)
Comment (jb2197) [reply MOPTools/twa_mops/ontospecies.py]: Addressed in https://github.com/cambridge-cares/TheWorldAvatar/pull/1330/commits/2b36b35b9b6d806f9bbabfc6bebf16ee822f2f51

**Review by sm453 — APPROVED** [review]

**Pull request merged**